### DISCUSSION TITLE
How to use DropDown with model?

### DISCUSSION THREAD

**Discussion by eminfedar** [Relm4/Relm4#512] [Q&A]
On gtk4-rs side, dropdown usage is really a mess. GObject inheriting complicates things and if I need to set a model to a dropdown like this, it is really a mess:
```rust
struct CountryName {
    country_id: i32,
    country_name: String
}
```

Basically the goal is: show a list of countries and when a country selected, get the selected country's id to call an api with it (like weather status).

On Pure GTK4 Side, I need to define a [PropertyExpression](https://docs.rs/gtk4/latest/gtk4/struct.PropertyExpression.html) and a [model](https://docs.rs/gtk4/latest/gtk4/struct.DropDown.html#method.set_model). And inherit the GObject stuff for my model struct to handle it.

Is there any example about how can I create a DropDown with a model and expression easily? Should I follow what I do exactly on pure gtk4 side? Is there an easy way to do it (like factory? _but I suppose it doesn't fit in this case?_)?

**Comment by 13hannes11** [comment]
I managed to get something working using the gtk4 book [example](https://gtk-rs.org/gtk4-rs/git/book/list_widgets.html). Maybe this is helpful.

```rust
    fn init(
        _: Self::Init,
        root: Self::Root,
        _sender: ComponentSender<Self>,
    ) -> ComponentParts<Self> {
        let model = Self {};
        let selection_model = gtk::gio::ListStore::new::<gtk::StringObject>();

        let sort_function =  | to_insert: &gtk::glib::Object, existing: &gtk::glib::Object | -> std::cmp::Ordering { to_insert.cmp(existing) };
        selection_model.insert_sorted(&gtk::StringObject::new("Test"), &sort_function);
        selection_model.insert_sorted(&gtk::StringObject::new("Test2"), &sort_function);
        selection_model.insert_sorted(&gtk::StringObject::new("Test3"), &sort_function);

        let selection = 2;

        view! {
            widgets = root.clone() {
                add = &adw::PreferencesPage::new() {
                    add = &adw::PreferencesGroup::new() {
                    set_title: "Terminal settings",
                        add = &adw::ComboRow::new() {
                            set_title: "Terminal",
                            #[wrap(Some)]
                            set_model = &gtk::SingleSelection::new(Some(selection_model)),
                            #[watch]
                            set_selected: selection,
                        },
                    },
                },
            }
        }
        ComponentParts { model, widgets }
    }
```
